crates/extensions/c8y_mapper_ext/src/availability/builder.rs

```diff
@@ -0,0 +1,85 @@
+use crate::availability::actor::AvailabilityActor;
+use crate::availability::AvailabilityConfig;
+use crate::availability::AvailabilityInput;
+use crate::availability::AvailabilityOutput;
+use crate::availability::TimerComplete;
+use crate::availability::TimerStart;
+use std::convert::Infallible;
+use tedge_actors::Builder;
+use tedge_actors::CloneSender;
+use tedge_actors::DynSender;
+use tedge_actors::LoggingSender;
+use tedge_actors::MessageSink;
+use tedge_actors::MessageSource;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::RuntimeRequestSink;
+use tedge_actors::Service;
+use tedge_actors::SimpleMessageBoxBuilder;
+use tedge_api::mqtt_topics::ChannelFilter;
+use tedge_api::mqtt_topics::EntityFilter;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::TopicFilter;
+
+pub struct AvailabilityBuilder {
+    config: AvailabilityConfig,
+    box_builder: SimpleMessageBoxBuilder<AvailabilityInput, AvailabilityOutput>,
+    mqtt_publisher: DynSender<MqttMessage>,
+    timer_sender: DynSender<TimerStart>,
+}
+
+impl AvailabilityBuilder {
+    pub fn new(
+        config: AvailabilityConfig,
+        mqtt: &mut (impl MessageSource<MqttMessage, TopicFilter> + MessageSink<MqttMessage>),
```

Comment:
In the context of the message race that you highlighted between the 117 message emitted by this actor and the 101 message emitted by the mapper actor, this might be the root of it: Both the actors connected directly to the broker and acting on them independently without any coordination. The previous iteration of this PR was getting around that coordination by making the mapper actor do everything. This is not an issue limited to this actor alone, but in future when we extract the operation managers also from the mapper actor as independent actors, we'll face the exact same issues in terms of race between operation capability messages and the device registration messages.

One solution to this problem is to have some coordination between these actors and the mapper actor. So, in this case, instead of this actor processing MQTT messages directly from the broker, let the mapper be the central actor that deals with the conversion and despatch of MQTT messages. And then these actors would only receive proper typed messages from the mapper actor. Similarly, instead of these actors publishing MQTT messages directly, let them route those via the mapper actor.

That way, we can avoid the possibility of any message races, as the same mapper actor can make sure that it publishes the registration message before the converted message is handed over to the availability actor and similarly the converted availability messages sent by this actor are also published by the same mapper actor ensuring that it always follows the registration message that was previously sent.

These in-memory channels(message boxes) between these actors and the mapper actor also enables them to query the main entity store maintained by the mapper, instead of each actor maintaining their own "lightweight" entity stores.

Reply:
Your proposal is definitely a proper way to address the issue. However, it will block the availability feature again. Can't we just put a delay (e.g. sleep 1s) before MqttMessage to the MQTT actor? Or having proper solution is the highest priority? I need a guide what to do next

Reply:
> One solution to this problem is to have some coordination between these actors and the mapper actor. So, in this case, instead of this actor processing MQTT messages directly from the broker, let the mapper be the central actor that deals with the conversion and despatch of MQTT messages. And then these actors would only receive proper typed messages from the mapper actor. Similarly, instead of these actors publishing MQTT messages directly, let them route those via the mapper actor.

@albinsuresh Your proposal is definitely the way to go.

> Your proposal is definitely a proper way to address the issue. However, it will block the availability feature again. Can't we just put a delay (e.g. sleep 1s) before MqttMessage to the MQTT actor? Or having proper solution is the highest priority? I need a guide what to do next

@rina23q I appreciate your concerns and I agree that deferring again the availability feature is not desirable.
However, we must avoid this PR adding more work in the near future.

So I propose to change the input and output types of the `AvailabilityActor` to be fully unrelated to MQTT.
i.e. to directly consume `EntityRegistrationMessage` and `tedge_api::HealthStatus`
and to produce `c8y_api::HealthStatus`

```
fan_in_message_type!(AvailabilityInput[EntityRegistrationMessage, tedge_ap::HealthStatus] : Debug);
fan_in_message_type!(AvailabilityOutput[[MqttMessage](c8y_api::HealthStatus)] : Debug);
```

MQTT message parsing, which is currently done in `AvailabilityActor::process_mqtt_message()`,
will have to be move in `AvailabilityBuilder::new()` using something along the lines of [`MessageSink::connect_mapped_source`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_actors/src/builders.rs#L161) or [`MessageSource::connect_mapped_sink()``](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_actors/src/builders.rs#L202).

This will prepare for the refactoring proposed by @albinsuresh where the mapping actor will be the interface between MQTT and in-memory channels forwarding `tedge_api::*` messages to the actual actors and receiving back `c8y_api::*` messages.

Reply:
See https://github.com/thin-edge/thin-edge.io/commit/4e1d3d58be403f12d5674c0575067bc7e191d97d for an example

Reply:
> This solution is not 100% fool-proof because there is a corner case where the `@id`, `@health` and `@name` are all specified in the registration message, the 117 message might create the child with the wrong "name".

I don't get that. Why an explicit  `@name` might lead to a message with the wrong name?

Reply:
@albinsuresh 
> Instead of a delay, I'd just make the availability actor react only to registration messages with an `@id` in it

It's already like that. Otherwise there is no way to know `external ID` for child devices :D Also, for the main device, the device registration is done during `tedge connect c8y`, so it's out of concern.

> This solution is not 100% fool-proof because there is a corner case where the `@id`, `@health` and `@name` are all specified in the registration message, the 117 message might create the child with the wrong "name".

Both `@name` and `@health` are unrelated to 117 messages. Just a corner case is where the `@id` is provided in the initial registration message for child devices, in the case, tedge-mapper-c8y doesn't republishes the message since already `@id` is there.

Reply:
ca3027605560c40fc0bfddb1585db7ee08199bf3

> See [4e1d3d5](https://github.com/thin-edge/thin-edge.io/commit/4e1d3d58be403f12d5674c0575067bc7e191d97d) for an example

ca3027605560c40fc0bfddb1585db7ee08199bf3 changed the code like the example above. The actor does no longer consume `MqttMessage` as input and output ;)

Reply:
Actually [robot tests failures](https://github.com/thin-edge/thin-edge.io/actions/runs/9585880674/job/26456761734) proves that the race condition of `101` and `117` really occurs :/ I'd try to add a small delay and how the result will be.

Reply:
Resolved by routing messages to the availability actor through the mapper actor.